---
title: "Differentially expressed RNA between brain regions_pheatmap"
author: "Yuting Xie"
date: "2022/07/25"
output: html_document
---
rm(list = ls())
pacman::p_load(readxl,magrittr,
               RColorBrewer,
               Rtsne, umap,
               pheatmap,vioplot,
               ggpubr, ggplot2,
               corrplot, stringr,reshape2,preprocessCore)


#data
```{r}
RNA_info<-data.table::fread("//172.16.13.136/share/members/xieyuting/HBA/differential analysis of brain regions_20230111/Info_Allen_mRNA_sub_matrix_9162prot_2113sample20220920.csv",data.table = F)

pro_info<-data.table::fread("//172.16.13.136/share/members/xieyuting/HBA/differential analysis of brain regions_20230111/FIleInfo_HBA_20220816_withoutFALSE_and_pool.csv",data.table = F)

RNA_info$Lobe<-pro_info$Lobe[match(RNA_info$region_matchID,pro_info$BBA_ID)]  

RNA_region<-unique(RNA_info$region_matchName)
pro_region<-unique(pro_info$BBA_Anatomical.and.modified.Cyto.architectonic.descriptions)

con_region<-intersect(RNA_region,pro_region)
con_region <- con_region[-which(con_region=="#N/A")]

setwd("//172.16.13.136/share/members/xieyuting/HBA/differential analysis of brain regions_20230111")
df<-load(file="mRNA2_matrix_row_gene_symbol_col_Well_id.RData")
df<-as.data.frame(t(mRNA2))
df1<-df

RNA_info$BBA_Lobe <- paste0(RNA_info$region_matchName, seq="--", RNA_info$Lobe)
df1$BBA_Lobe<-RNA_info$BBA_Lobe[match(rownames(df1),RNA_info$well_id)]

#delet Lobe$"Cerebellum","SubcorticalNuclei","NA"   ("InsularLobe" don't delet)
df2<-df1[-which(is.na(df1$BBA_Lobe)),]
# df3<-df2[-grep("Cerebellum",df2$Lobe),]
# df4<-df3[-grep("SubcorticalNuclei",df3$Lobe),]
df4<-df2

df5<-df4[-which(df4$BBA_Lobe=="#N/A--NA"),]

```

#Fold change
```{r}
region <- unique(df5$BBA_Lobe)

df6<-reshape2::melt(df5,id="BBA_Lobe")
df7<-reshape2::dcast(df6, BBA_Lobe~variable, mean, na.rm=T)
df7 <- data.frame(df7, row.names = "BBA_Lobe")

library(foreach)
library(doParallel)
cl <- makeCluster(40)
registerDoParallel(cl)

FC<-foreach(i=1:length(region),.combine=rbind) %:%
    foreach(j=1:length(region),.combine=rbind) %dopar% {
    
    data1<-df7[which(rownames(df7)==region[i]),]
    data2<-df7[which(rownames(df7)==region[j]),]
    
    # data_mean1 <- apply(data1,2,function(x){mean(as.numeric(x))})
    # data_mean2 <- apply(data2,2,function(x){mean(as.numeric(x))})
    
    data_FC <- as.data.frame(data1-data2)
    
    rownames(data_FC)<-paste(region[i],region[j],sep="_X_")
    
    return(data_FC)
  }

pacman::p_load(magrittr)
abs_log2_FC<-abs(FC) %>% as.data.frame()

stopImplicitCluster()
stopCluster(cl)

setwd("//172.16.13.136/share/members/xieyuting/HBA/differential analysis of brain regions_20230111")
save(abs_log2_FC, file="RNA_G_matter_abs_log2_FC.RData")
```

#t-test
```{r}
library(foreach)
library(doParallel)
cl <- makeCluster(40)
registerDoParallel(cl)

#df6 是求过对数后的data.frame

t<-foreach(i=1:length(region),.combine=rbind) %:%
        foreach(j=1:length(region),.combine=rbind) %dopar% {
    
        t1<-df5[c(which(df5$BBA_Lobe == region[i]),which(df5$BBA_Lobe==region[j])),]
        
        pval<-sapply(c(1:(ncol(t1)-1)),function(x){
              p<-try(t.test(t1[,x]~t1$BBA_Lobe),silent=TRUE)
              if(is(p,"try-error")) {NA} else {p$p.value}
              })
        
        t2 <- t(as.data.frame(pval))
        rownames(t2)<-paste(region[i],region[j],sep="_X_")
        
        return(t2)
  }

t<-as.data.frame(t)
colnames(t)<-colnames(df5)[-ncol(df5)]

adj_p<-as.data.frame(t(apply(t,1,function(x){p.adjust(x,method = "BH")})))

stopImplicitCluster()
stopCluster(cl)

setwd("//172.16.13.136/share/members/xieyuting/HBA/differential analysis of brain regions_20230111")
save(adj_p, file="RNA_G_matter_adj_p.RData")
```

#Selectproteins:  BH-corrected-P < 0.05
```{r}
setwd("Z:/members/xieyuting/HBA/differential analysis of brain regions_20230111")
load(file="RNA_G_matter_abs_log2_FC.RData")
load(file="RNA_G_matter_adj_p.RData")

library(foreach)
library(doParallel)
cl <- makeCluster(40)
registerDoParallel(cl)

#log2(FC) > 1.5 | log2(FC) < -1.5
setwd("//172.16.13.136/share/members/xieyuting/HBA/differential analysis of brain regions_20230111")
df<-load(file="mRNA2_matrix_row_gene_symbol_col_Well_id.RData")
df<-as.data.frame(t(mRNA2))
#计算过程中将abs_log2_FC中列名中"_"变为了".", 导致transcripts名字不一致
colnames(abs_log2_FC) <- colnames(df)
adj_p1 <- adj_p[rownames(abs_log2_FC),colnames(abs_log2_FC)]
adj_p1[is.na(adj_p1)]<-2

DP<-foreach(i=1:nrow(adj_p1),.combine=rbind,.inorder=TRUE) %dopar% {
  a<-adj_p1[i,]
  b<-abs_log2_FC[i,]
  data<-c()
  data[which((a < 0.05) & (b > 1.5))]<- 1
  data[which(a >= 0.05 | b <= 1.5)]<- 0   #log2(FC) < -1.5
  return(data)
}

DP<-as.data.frame(DP)
rownames(DP)<-rownames(adj_p1)
colnames(DP)<-colnames(adj_p1)


stopImplicitCluster()
stopCluster(cl)

# setwd("Z:/project/HBA_xiaoqi/DataAnalysis20220712/differential analysis of brain regions based on dysregulated proteins")
# write.csv(DP,"differential_protiens_abs_log2_FC_1.5_adj_P_0.05_no_NA.csv")
```

#heatmap
```{r}
HP<-as.data.frame(apply(DP,1,function(x){sum(x==1)}))
colnames(HP)<-"RNA_number"
HP$V1<-rownames(HP)
HP$V2<-rownames(HP)
HP$V1<- sapply(HP$V1,function(x){str_split(x,"_X_")[[1]][1]})
HP$V2<- sapply(HP$V2,function(x){str_split(x,"_X_")[[1]][2]})
HP$value<-HP$RNA_number
HP<-HP[,-1]
HP_data<-acast(HP,V1~V2)

# Na.ratio$a<-rep(1,98)
# Na.ratio1<-Na.ratio[rownames(HP_data),]
# rownames(HP_data)<-paste(colnames(HP_data),Na.ratio1$ratio_Na,sep="--NA_ratio-")

# setwd("W:/members/xieyuting/HBA/differential analysis of brain regions_20230111")
# save(DP, HP, HP_data, file="RNA_G_matter_differentially expressed RNA data_DP_HP_HP_data.RData")

setwd("D:/1_guomics_xyt/1Project/hba_research_workflow/9_DataAnalysis_20220901/differential analysis of brain regions_20230111")
load(file="RNA_G_matter_differentially expressed RNA data_DP_HP_HP_data.RData")
# info<-read.csv("FIleInfo_HBA_20220816_withoutFALSE_and_pool.csv",header = T,sep = ",",stringsAsFactors = F)
# info$Lobe_BBA <- paste0(info$BBA_Anatomical.and.modified.Cyto.architectonic.descriptions, seq="--", info$Lobe)
# info_order <- data.frame(BBA=info$BBA_Anatomical.and.modified.Cyto.architectonic.descriptions, Lobe_BBA=info$Lobe_BBA, order=info$Order, gyrus=info$Gyrus)
# info_order <- unique(info_order)
# info_order <- info_order[-which(info_order$BBA=="#N/A"), ]
# info_order <- info_order[order(info_order$order), ]
info_order <- read.csv("info_order_of_BBA.csv")

# names <- data.frame(name=rownames(HP_data))
# names$v1 <- sapply(names$name,function(x){str_split(x,"_")[[1]][1]})
# names$v2 <- sapply(names$name,function(x){str_split(x,"_")[[1]][2]})
# names$v3 <- sapply(names$name,function(x){str_split(x,"_")[[1]][3]})
# names$v4 <- sapply(names$name,function(x){str_split(x,"_")[[1]][4]})
# # names2 <- names[!is.na(names$v3), ]
# names$v5 <- paste0(names$v2, seq="_", names$v3, seq="_", names$v4)
# names$v6 <- sapply(names$v5,function(x){gsub("_NA","", x)})
# names$BBA_Lobe <- paste0(names$v6, seq="--", names$v1)

# rownames(HP_data) <- names$BBA_Lobe[match(rownames(HP_data), names$name)]
# colnames(HP_data) <- rownames(HP_data)
info_order1 <- info_order[match(rownames(HP_data), info_order$Lobe_BBA), ]
info_order1 <- info_order1[order(info_order1$order), ]
HP_data <- HP_data[info_order1$Lobe_BBA, info_order1$Lobe_BBA]
HP_data[HP_data==0]<-NA
# HP_data<-HP_data+1
log_HPdata<-log2(HP_data)
log_HPdata[is.na(log_HPdata)]<-0

rownames(log_HPdata) <- info_order1$ABB[match(rownames(log_HPdata), info_order1$Lobe_BBA)]
colnames(log_HPdata) <- rownames(log_HPdata)
# rownames(HP_data) <- fct_inorder(as.factor(rownames(HP_data)))
# colnames(HP_data) <- fct_inorder(as.factor(colnames(HP_data)))

anno <- data.frame(Lobe=info_order1$Lobe, gyrus=info_order1$gyrus)
rownames(anno) <- info_order1$ABB

setwd("D:/1_guomics_xyt/1Project/hba_research_workflow/9_DataAnalysis_20220901/differential analysis of brain regions_20230111")
colors <- read.csv("colors.csv")
colors <- colors[, -1]

lobe_col <- colors[match(unique(anno$Lobe), colors$group), ]
lobe_col1 <- lobe_col$colors
names(lobe_col1) <- lobe_col$group

gyrus_col <- colors[match(unique(anno$gyrus), colors$group), ]
gyrus_col1 <- gyrus_col$colors
names(gyrus_col1) <- gyrus_col$group

ann_colors = list( Lobe=lobe_col1, gyrus=gyrus_col1)

setwd("D:/1_guomics_xyt/1Project/hba_research_workflow/9_DataAnalysis_20220901/differential analysis of brain regions_20230111")
library(pheatmap)
pheatmap(log_HPdata,
         fontsize=15,
         color = colorRampPalette(c("white","#6A7F67","firebrick4"))(500), 
         #annotation_col = colnames(pathway2),
         #annotation_colors = list(W_G=c(rep("#4472C4",26),rep("#EE781C",19),Region=rainbow(45))),
         #na_color = "gray",
         annotation_col = anno,
         annotation_colors =  ann_colors,
         # annotation_row = lobe,
         # annotation_row = anno,
         trace = "none", 
         scale = "none", 
         cluster_rows = F,
         cluster_cols = F,
         show_rownames = T, 
         show_colnames = T, 
         filename = "RNA_G_matter_differential_protein_numbers_AbsLog2FC_1.5_114_Regions_annotation.pdf",       
         main = "Number of differentially expressed RNA between pairs of brain regions in grey matter",
         height = 40,width =44)








# HP_order<-as.data.frame(HP_data)
# HP_order$lobe<-rownames(HP_order)
# HP_order$lobe<-sapply(HP_order$lobe,function(x){str_split(x,"---")[[1]][2]})
# 
# HP_order1<-HP_order[order(HP_order$lobe,decreasing=T),-ncol(HP_order)] 
# HP_order2<-HP_order1[,rownames(HP_order1)]
# 
# annotation_col=data.frame(HP_order$lobe)
# rownames(annotation_col)<-rownames(HP_order)
# 
# 
# setwd("Z:/project/HBA_xiaoqi/DataAnalysis20220712/differential analysis of brain regions_20220722")
# pheatmap(HP_order2,
#          fontsize=15,
#          Colv=NA,Rowv=NA, #取消树状图
#          color = colorRampPalette(c("white","#6A7F67","firebrick4"))(500), 
#          annotation_col = annotation_col,
#          annotation_row = annotation_col,
#          #annotation_colors = list(W_G=c(rep("#4472C4",26),rep("#EE781C",19),Region=rainbow(45))),
#          #na_color = "gray",
#          trace = "none", 
#          scale = "none",
#          cluster_rows = F,
#          cluster_cols = F,
#          #clustering_method="median", #"ward" ward.D" "centroid" "ward.D2" "single" "complete" "average" "mequitty"
#          #clustering_distance_rows="correlaition",clustering_distance_cols="correlaition", #pearson
#          show_rownames = T, 
#          show_colnames = T, 
#          #angle_col=45,angle_row=45,#字体角度
#          height = 40,width =40,
#          xlab="Brain Regions", ylab="Brain Regions",
#          main="Number of differentially expressed genes between pairs of stuctures",
#          #cutree_rows=10, cutree_cols=10,
#          #gaps_row=c(), gaps_col=c(),
#          #treeheight_row=100,treeheight_col=100,
#          #cellwidth=10,cellheight=10,
#          #display_numbers=TRUE, number_color="black",
#          display_numbers=ifelse(HP_order2>300,"*",""),
#          #border_color="blak" #border=FALSE,
#          #lengend_breaks=c(0:2000),legend_labels=as.character(seq(from=0,to=2000,by=200)),
#          #lengend=FALSE,
#          filename = "Number of differentially expressed genes between pairs of stuctures_lobe_order.pdf"
# )






```

```{r}
stopImplicitCluster()
stopCluster(cl)
```
